<a href="https://colab.research.google.com/github/YoungJang-KR/semiconductor-fault-detection/blob/main/02_merge_sampling_PHM18_etching.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# 라이브러리 불러오기

import pandas as pd
import os
import numpy as np
import glob

In [3]:
dir_train = '/content/drive/MyDrive/phm18/train'
dir_ttf = '/content/drive/MyDrive/phm18/train_ttf'

files_train = sorted( glob.glob(os.path.join(dir_train, '*csv')) ) # '*csv'를 통해 확장자가 .csv인 파일만 골라내기
files_ttf = sorted( glob.glob(os.path.join(dir_ttf, '*csv')) )

print(f"train 파일 개수: {len(files_train)}")
print(files_train)
print()
print(f"train_ttf 파일 개수: {len(files_ttf)}")
print(files_ttf)

train 파일 개수: 20
['/content/drive/MyDrive/phm18/train/01_M01_DC_train.csv', '/content/drive/MyDrive/phm18/train/01_M02_DC_train.csv', '/content/drive/MyDrive/phm18/train/02_M01_DC_train.csv', '/content/drive/MyDrive/phm18/train/02_M02_DC_train.csv', '/content/drive/MyDrive/phm18/train/03_M01_DC_train.csv', '/content/drive/MyDrive/phm18/train/03_M02_DC_train.csv', '/content/drive/MyDrive/phm18/train/04_M01_DC_train.csv', '/content/drive/MyDrive/phm18/train/04_M02_DC_train.csv', '/content/drive/MyDrive/phm18/train/05_M01_DC_train.csv', '/content/drive/MyDrive/phm18/train/05_M02_DC_train.csv', '/content/drive/MyDrive/phm18/train/06_M01_DC_train.csv', '/content/drive/MyDrive/phm18/train/06_M02_DC_train.csv', '/content/drive/MyDrive/phm18/train/07_M01_DC_train.csv', '/content/drive/MyDrive/phm18/train/07_M02_DC_train.csv', '/content/drive/MyDrive/phm18/train/08_M01_DC_train.csv', '/content/drive/MyDrive/phm18/train/08_M02_DC_train.csv', '/content/drive/MyDrive/phm18/train/09_M01_DC_train.csv

In [4]:
# 컬럼 정의하기
sensor_cols = [
    'IONGAUGEPRESSURE', 'ETCHSUPPRESSORCURRENT', 'FLOWCOOLPRESSURE',
    'ACTUALROTATIONANGLE', 'ACTUALSTEPDURATION'
]

control_cols = [
    'ETCHBEAMVOLTAGE', 'ETCHBEAMCURRENT', 'ETCHSUPPRESSORVOLTAGE',
    'FLOWCOOLFLOWRATE', 'ETCHGASCHANNEL1READBACK', 'ETCHPBNGASREADBACK',
    'FIXTURETILTANGLE', 'ROTATIONSPEED', 'FIXTURESHUTTERPOSITION',
    'ETCHSOURCEUSAGE', 'ETCHAUXSOURCETIMER', 'ETCHAUX2SOURCETIMER'
]

ttf_cols = [
    'TTF_pressure_drop', 'TTF_pressure_high', 'TTF_leak'
]

In [5]:
# 데이터 확인

df_01 = pd.read_csv(files_train[0])
print(df_01.shape)
print(df_01.columns.tolist())

(3124704, 24)
['time', 'Tool', 'stage', 'Lot', 'runnum', 'recipe', 'recipe_step', 'IONGAUGEPRESSURE', 'ETCHBEAMVOLTAGE', 'ETCHBEAMCURRENT', 'ETCHSUPPRESSORVOLTAGE', 'ETCHSUPPRESSORCURRENT', 'FLOWCOOLFLOWRATE', 'FLOWCOOLPRESSURE', 'ETCHGASCHANNEL1READBACK', 'ETCHPBNGASREADBACK', 'FIXTURETILTANGLE', 'ROTATIONSPEED', 'ACTUALROTATIONANGLE', 'FIXTURESHUTTERPOSITION', 'ETCHSOURCEUSAGE', 'ETCHAUXSOURCETIMER', 'ETCHAUX2SOURCETIMER', 'ACTUALSTEPDURATION']


In [6]:
dfs = [] # 총 20개의 데이터를 담을 바구니 만들기

for file_train, file_ttf in zip(files_train, files_ttf):
  df_train = pd.read_csv(file_train)
  df_ttf = pd.read_csv(file_ttf)
  df_merged = df_train.merge(df_ttf, on='time')
  df_sampled = df_merged.sample(frac=0.01, random_state=42) # 합친 데이터에서 1%만 랜덤 추출, random_state (중요)
  dfs.append(df_sampled) # sampled 결과를 바구니에 넣기
  print(f"{file_train} 처리 완료: {df_sampled.shape}")

df_all_1pct = pd.concat(dfs, axis=0, ignore_index=True) # ignore_index=True는 행 번호를 0부터 새로 매겨줌
print(df_all_1pct.shape)

/content/drive/MyDrive/phm18/train/01_M01_DC_train.csv 처리 완료: (31357, 27)
/content/drive/MyDrive/phm18/train/01_M02_DC_train.csv 처리 완료: (51260, 27)
/content/drive/MyDrive/phm18/train/02_M01_DC_train.csv 처리 완료: (28689, 27)
/content/drive/MyDrive/phm18/train/02_M02_DC_train.csv 처리 완료: (41489, 27)
/content/drive/MyDrive/phm18/train/03_M01_DC_train.csv 처리 완료: (34344, 27)
/content/drive/MyDrive/phm18/train/03_M02_DC_train.csv 처리 완료: (50630, 27)
/content/drive/MyDrive/phm18/train/04_M01_DC_train.csv 처리 완료: (51013, 27)
/content/drive/MyDrive/phm18/train/04_M02_DC_train.csv 처리 완료: (44529, 27)
/content/drive/MyDrive/phm18/train/05_M01_DC_train.csv 처리 완료: (48204, 27)
/content/drive/MyDrive/phm18/train/05_M02_DC_train.csv 처리 완료: (19262, 27)
/content/drive/MyDrive/phm18/train/06_M01_DC_train.csv 처리 완료: (38176, 27)
/content/drive/MyDrive/phm18/train/06_M02_DC_train.csv 처리 완료: (51075, 27)
/content/drive/MyDrive/phm18/train/07_M01_DC_train.csv 처리 완료: (53232, 27)
/content/drive/MyDrive/phm18/train/07_

In [7]:
# 매번 샘플링하기 번거로우므로, 샘플링 된 결과를 저장해두자

df_all_1pct.to_csv('/content/drive/MyDrive/phm18/df_all_sampled_1pct.csv', index=False)
# index=False는 나중에 데이터를 불러올 때 불필요한 컬럼이 생기지 않게 해줌
print("저장 완료!", df_all_1pct.shape)

저장 완료! (824313, 27)


In [9]:
# 불러오기
df_all_1pct = pd.read_csv('/content/drive/MyDrive/phm18/df_all_sampled_1pct.csv')
print("행, 렬:", df_all_1pct.shape)
print("열 항목:", df_all_1pct.columns.tolist())

행, 렬: (824313, 27)
열 항목: ['time', 'Tool', 'stage', 'Lot', 'runnum', 'recipe', 'recipe_step', 'IONGAUGEPRESSURE', 'ETCHBEAMVOLTAGE', 'ETCHBEAMCURRENT', 'ETCHSUPPRESSORVOLTAGE', 'ETCHSUPPRESSORCURRENT', 'FLOWCOOLFLOWRATE', 'FLOWCOOLPRESSURE', 'ETCHGASCHANNEL1READBACK', 'ETCHPBNGASREADBACK', 'FIXTURETILTANGLE', 'ROTATIONSPEED', 'ACTUALROTATIONANGLE', 'FIXTURESHUTTERPOSITION', 'ETCHSOURCEUSAGE', 'ETCHAUXSOURCETIMER', 'ETCHAUX2SOURCETIMER', 'ACTUALSTEPDURATION', 'TTF_FlowCool Pressure Dropped Below Limit', 'TTF_Flowcool Pressure Too High Check Flowcool Pump', 'TTF_Flowcool leak']


In [11]:
# 열 항목 이름 줄이기
df_all_1pct = df_all_1pct.rename( columns={'TTF_FlowCool Pressure Dropped Below Limit': 'TTF_pressure_drop',
                                           'TTF_Flowcool Pressure Too High Check Flowcool Pump': 'TTF_pressure_high',
                                           'TTF_Flowcool leak': 'TTF_leak'

})

# 확인
print("열 항목 확인:", df_all_1pct.columns.tolist())

열 항목 확인: ['time', 'Tool', 'stage', 'Lot', 'runnum', 'recipe', 'recipe_step', 'IONGAUGEPRESSURE', 'ETCHBEAMVOLTAGE', 'ETCHBEAMCURRENT', 'ETCHSUPPRESSORVOLTAGE', 'ETCHSUPPRESSORCURRENT', 'FLOWCOOLFLOWRATE', 'FLOWCOOLPRESSURE', 'ETCHGASCHANNEL1READBACK', 'ETCHPBNGASREADBACK', 'FIXTURETILTANGLE', 'ROTATIONSPEED', 'ACTUALROTATIONANGLE', 'FIXTURESHUTTERPOSITION', 'ETCHSOURCEUSAGE', 'ETCHAUXSOURCETIMER', 'ETCHAUX2SOURCETIMER', 'ACTUALSTEPDURATION', 'TTF_pressure_drop', 'TTF_pressure_high', 'TTF_leak']


In [12]:
# 바뀐 이름으로 새로이 저장
df_all_1pct.to_csv('/content/drive/MyDrive/phm18/df_all_sampled_1pct.csv', index=False)
print("저장 완료!")

저장 완료!


In [13]:
# 결측값 확인
df_all_1pct = pd.read_csv('/content/drive/MyDrive/phm18/df_all_sampled_1pct.csv')

print("행, 렬:", df_all_1pct.shape)
print("열 항목:", df_all_1pct.columns.tolist())
print("결측값", df_all_1pct.isna().sum())

행, 렬: (824313, 27)
열 항목: ['time', 'Tool', 'stage', 'Lot', 'runnum', 'recipe', 'recipe_step', 'IONGAUGEPRESSURE', 'ETCHBEAMVOLTAGE', 'ETCHBEAMCURRENT', 'ETCHSUPPRESSORVOLTAGE', 'ETCHSUPPRESSORCURRENT', 'FLOWCOOLFLOWRATE', 'FLOWCOOLPRESSURE', 'ETCHGASCHANNEL1READBACK', 'ETCHPBNGASREADBACK', 'FIXTURETILTANGLE', 'ROTATIONSPEED', 'ACTUALROTATIONANGLE', 'FIXTURESHUTTERPOSITION', 'ETCHSOURCEUSAGE', 'ETCHAUXSOURCETIMER', 'ETCHAUX2SOURCETIMER', 'ACTUALSTEPDURATION', 'TTF_pressure_drop', 'TTF_pressure_high', 'TTF_leak']
결측값 time                            0
Tool                            0
stage                           0
Lot                             0
runnum                          0
recipe                          0
recipe_step                     8
IONGAUGEPRESSURE                0
ETCHBEAMVOLTAGE                 0
ETCHBEAMCURRENT                 0
ETCHSUPPRESSORVOLTAGE           0
ETCHSUPPRESSORCURRENT           0
FLOWCOOLFLOWRATE                0
FLOWCOOLPRESSURE                0
ETCH

In [16]:
threshold = 1839644 # 01_M01_DC_train.csv 파일 분석 때와 동일한 threshold 사용

df_all_1pct['label_pressure_high'] = ( df_all_1pct['TTF_pressure_high'] <= threshold ).astype(int)
print(df_all_1pct['label_pressure_high'].value_counts(normalize=True)) # 수치로 보여줌

label_pressure_high
0    0.763006
1    0.236994
Name: proportion, dtype: float64


전체 파일에 대한 1% 샘플링에서도 '01_M01_DC_train.csv' 파일 때와 비슷하게 고장 임박 구간이 전체의 23.7%를 유지한다.